### PR TITLE
Bumpup bmc-log-collector  0.9.1  fix bugs 

### PR DESCRIPTION
**Pull request by takara9**

### CODE REVIEW

**Review by masa213f — CHANGES_REQUESTED** [review]

File: bmc-log-collector/log-collector.go [hunk 1]
```diff
@@ -90,66 +89,84 @@ func (c *selCollector) collectSystemEventLog(ctx context.Context, m Machine, log
...
-
-		if lastPtr.LastReadId < lastId {
-			// Normal case without iDRAC log reset.
-			bmcByteJsonLog, err := json.Marshal(members.Sel[i])
+	createTime, err := time.Parse(layout, response.Sel[len(response.Sel)-1].Create)
+	if err != nil {
+		slog.Error("failed to parse for time", "err", err, "serial", m.Serial, "currentLastReadId", currentLastId, "ptrDir", c.ptrDir)
+		return
+	}
+	FirstCreateTime := createTime.Unix()
```
[the remark this answers is not in the document]
Comment (takara9) [reply bmc-log-collector/log-collector.go:111]: updated

File: bmc-log-collector/log-collector.go [hunk 2]
```diff
@@ -55,8 +55,7 @@ type selCollector struct {
 
 func (c *selCollector) collectSystemEventLog(ctx context.Context, m Machine, logWriter bmcLogWriter) {
 	const layout = "2006-01-02T15:04:05Z07:00"
```
[the remark this answers is not in the document]
Comment (takara9) [reply bmc-log-collector/log-collector.go:57]: Updated.

File: bmc-log-collector/log-collector.go [hunk 3]
```diff
@@ -90,66 +89,84 @@ func (c *selCollector) collectSystemEventLog(ctx context.Context, m Machine, log
...
+	}
+	FirstCreateTime := createTime.Unix()
+
+	for i := len(response.Sel) - 1; i >= 0; i-- {
+		t, err := time.Parse(layout, response.Sel[i].Create)
+		if err != nil {
+			slog.Error("failed to time parse", "err", err, "serial", m.Serial, "LastReadId", currentLastId, "ptrDir", c.ptrDir)
+			continue
+		}
+		currentCreateUnixtime := t.Unix()
```
[the remark this answers is not in the document]
Comment (takara9) [reply bmc-log-collector/log-collector.go:119]: eleminated the redundancy code.

File: bmc-log-collector/log-collector.go [hunk 4]
```diff
@@ -55,8 +55,7 @@ type selCollector struct {
 
 func (c *selCollector) collectSystemEventLog(ctx context.Context, m Machine, logWriter bmcLogWriter) {
 	const layout = "2006-01-02T15:04:05Z07:00"
-	var createUnixtime int64
-	var lastId int
+	var currentLastId int
```
[the remark this answers is not in the document]
Comment (takara9) [reply bmc-log-collector/log-collector.go:58]: deleted.

File: bmc-log-collector/log-collector.go [hunk 5]
```diff
@@ -90,66 +89,84 @@ func (c *selCollector) collectSystemEventLog(ctx context.Context, m Machine, log
 
 		err = updateLastPointer(lastPtr, filePath)
 		if err != nil {
-			slog.Error("failed to write a pointer file.", "err", err, "serial", m.Serial, "createUnixtime", createUnixtime, "LastReadId", lastId, "ptrDir", c.ptrDir)
+			slog.Error("failed to write a pointer file.", "err", err, "serial", m.Serial, "currentLastReadId", currentLastId, "ptrDir", c.ptrDir)
```
[the remark this answers is not in the document]
Comment (takara9) [reply bmc-log-collector/log-collector.go:92]: deleted.

File: bmc-log-collector/log-collector.go [hunk 6]
```diff
@@ -90,66 +89,84 @@ func (c *selCollector) collectSystemEventLog(ctx context.Context, m Machine, log
...
-		members.Sel[i].Serial = m.Serial
-		members.Sel[i].BmcIP = m.BmcIP
-		members.Sel[i].NodeIP = m.NodeIP
-
-		if lastPtr.LastReadId < lastId {
-			// Normal case without iDRAC log reset.
-			bmcByteJsonLog, err := json.Marshal(members.Sel[i])
+	createTime, err := time.Parse(layout, response.Sel[len(response.Sel)-1].Create)
+	if err != nil {
+		slog.Error("failed to parse for time", "err", err, "serial", m.Serial, "currentLastReadId", currentLastId, "ptrDir", c.ptrDir)
```
[the remark this answers is not in the document]
Comment (takara9) [reply bmc-log-collector/log-collector.go:108]: changed.

File: bmc-log-collector/log-collector.go [hunk 7]
```diff
@@ -90,66 +89,84 @@ func (c *selCollector) collectSystemEventLog(ctx context.Context, m Machine, log
...
+	FirstCreateTime := createTime.Unix()
+
+	for i := len(response.Sel) - 1; i >= 0; i-- {
+		t, err := time.Parse(layout, response.Sel[i].Create)
+		if err != nil {
+			slog.Error("failed to time parse", "err", err, "serial", m.Serial, "LastReadId", currentLastId, "ptrDir", c.ptrDir)
+			continue
+		}
+		currentCreateUnixtime := t.Unix()
+		currentLastId, err = strconv.Atoi(response.Sel[i].Id)
```
[the remark this answers is not in the document]
Comment (takara9) [reply bmc-log-collector/log-collector.go:120]: changed

File: bmc-log-collector/log-collector.go [hunk 8]
```diff
@@ -90,66 +89,84 @@ func (c *selCollector) collectSystemEventLog(ctx context.Context, m Machine, log
...
-				slog.Error("failed to output log", "err", err, "serial", m.Serial, "createUnixtime", createUnixtime, "LastReadId", lastId, "ptrDir", c.ptrDir)
+				slog.Error("failed to output log", "err", err, "serial", m.Serial, "bmcByteJsonLog", string(bmcByteJsonLog), "currentLastReadId", currentLastId, "ptrDir", c.ptrDir)
+			}
+
+			lastPtr.LastReadId = currentLastId
+			lastPtr.LastReadTime = currentCreateUnixtime
+			lastPtr.LastError = nil
+			if currentLastId == 1 {
+				lastPtr.FirstCreateTime = currentCreateUnixtime
 			}
```
[the remark this answers is not in the document]
Comment (takara9) [reply bmc-log-collector/log-collector.go:146]: deleted.

File: bmc-log-collector/log-collector.go [hunk 9]
```diff
@@ -90,66 +87,75 @@ func (c *selCollector) collectSystemEventLog(ctx context.Context, m Machine, log
...
 			err = logWriter.write(string(bmcByteJsonLog), m.Serial)
 			if err != nil {
-				slog.Error("failed to output log", "err", err, "serial", m.Serial, "createUnixtime", createUnixtime, "LastReadId", lastId, "ptrDir", c.ptrDir)
+				slog.Error("failed to output log", "err", err, "serial", m.Serial, "bmcByteJsonLog", string(bmcByteJsonLog), "currentLastReadId", currentId, "ptrDir", c.ptrDir)
 			}
-			lastPtr.LastReadId = lastId
-			lastPtr.LastReadTime = createUnixtime
+
+			lastPtr.LastReadId = currentId
+			lastPtr.LastReadTime = firstCreateTime
```
Comment (masa213f) [inline bmc-log-collector/log-collector.go:134]: Sorry for the additional comment.
`LastReadTime` is no longer needed, isn't it?

Comment (takara9) [reply bmc-log-collector/log-collector.go:134]: sure, it is not using variable.
deleted

**Review by masa213f — APPROVED** [review]
LGTM

**Pull request merged**